Add Basic Transform Functionality

kahnvex opened the pull request:
This adds basic transform functionality to browserify-i18n, as well as an optimized function which runs much faster, but is called in a more awkward way. See the README file for usage instructions.

This also adds tests validating the output of the translation replacement.

@imorrison 

Edited:

For now, this only works for `.hbs` files that use the `{{trans ".."}}` tag. This should obviously be made configurable, but I want to do this iteratively, and get it working first.

kahnvex committed:
Add application boilerplate to browserify-i18n

kahnvex committed:
Add application code and tests to browserify i18n

This adds basic functionality, as well as an optimized function that can be used to run browserify-i18n. This also adds tests validating the output of the translation replacement.

kahnvex committed:
Missing newline at EOF

kahnvex committed:
Add a .lintrc file.

kahnjw commented:
@beck is probably also interested.

@thomasw FYI

kahnvex committed:
Add test instructions to README

imorrison commented on README.md:
````diff
@@ -1,2 +1,56 @@
+## Test
+
+```sh
+$ mocha
````
`npm test` ?

beck replied:
I prefer just `mocha`, it gives confidence that the tests are setup with familiar convention.

`npm test` means I have to dig in package.json to see what's really happening

imorrison replied:
running `npm` means that the task it scoped to the app.... but I am not particularly opinionated about this.

kahnvex replied:
`npm test` still works. I agree with @imorrison because you have to have `mocha` globally installed for this to work. But I am not very opinionated about this.

imorrison commented on index.js:
```diff
@@ -0,0 +1,80 @@
+  var template = Handlebars.compile(chunk.toString());
+  var chunkString = '';
+
+  chunkString += 'module.exports = "';
```
This is odd but it does make sense.

imorrison commented:
I think this is solid. You should factor out the Handlebars dependency, which it sounds like you are going to do in another PR. 

---

I am not going to give this a thumbs yet, so @beck has time to also look at this.

beck commented on index.js:
```diff
@@ -0,0 +1,80 @@
+  var chunkString = '';
+
+  chunkString += 'module.exports = "';
+  chunkString += jsesc(template(), {quotes: 'double'});
```
unfamiliar with `jsesc`, the docs say:

> generating the shortest possible valid ASCII-only output

ASCII-only seems like any msgids with a non-ascii char will not get the right translation.

kahnvex replied:
Shoot, that is probably correct. I'll add a test for this.

kahnvex replied:
Yep, its failing.

kahnvex replied:
Fixed it. Let me know if you think my solution will work, its fairly simple.

beck commented:
This looks pretty cool.

I'm unable to get the tests to pass, [the output](https://gist.github.com/beck/fb2f309640db50d9fc06).

beck commented on README.md:
````diff
@@ -1,2 +1,56 @@
+For standard, non-optimized usage, do:
+
+```javascript
+var browserify = require('browserif');
````
s/browserif/browserify

kahnjw commented:
>I'm unable to get the tests to pass, the output.

@euwest and I are debugging this atm. Sorry, this commit 5937cd7  breaks it. It is an escaping issue.

kahnvex committed:
Validate that translation adds a double escaped \n

kahnvex committed:
Use a custom lodash template interpolator

kahnjw commented:
@beck should be fixed.

kahnvex committed:
Fix typo

kahnvex committed:
Assert it replaces text with special chars

kahnvex committed:
Don't use jsesc, it escapes special characters

kahnvex committed:
Handlebars setup function is misnamed

It no longer does anything with handlebars, it simply returns the catalog.

euwest commented on index.js:
```diff
@@ -0,0 +1,84 @@
+  var translatedString = template(catalog)
+    .replace(/\n/g, '\\n')
+    .replace(/"/g, '\"');
+
```
``` js
.replace(/(\n|")/g, '\\$&');
```

this may also need to include some other things besides just new line chars and double quotes

kahnvex committed:
Add option for custom trans tag interpolator

kahnvex committed:
Add documentation on using a custom interpolator

kahnvex committed:
Merge options hashes in optimized translate transform

kahnvex committed:
Improve interpolation documentation

kahnvex committed:
More interpolation docs improvement

beck commented on index.js:
```diff
@@ -0,0 +1,84 @@
+
+
+var replaceText = function(catalog, opts, chunk, enc, callback) {
+  var template = _.template(chunk.toString(), {
```
Why the switch from handlebars to lodash templates?

I'm not familiar with the difference between the two but if the feature set is not identical then this seems like a source for potential issue.

I've learned the hard way that sending whole files to lodash template is terribly inefficient—if the file has any size, it will exceed node's default stack size (and then you have to have nasty helper scripts to bump node's stack size).  I fully regret using it and advise not doing so here.

kahnvex replied:
> I've learned the hard way that sending whole files to lodash template is terribly inefficient—if the file has any size, it will exceed node's default stack size (and then you have to have nasty helper scripts to bump node's stack size). I fully regret using it and advise not doing so here.

When we ran it, it executed much faster than Handlebars. Of course that doesn't say much about memory consumption, or long call stacks. The reason we used lodash templates is that they allow us to use a custom interpolator, which is definitely necessary for consumers that want to bundle handlebars templates in their code which are not static.

kahnvex replied:
If we run into a situation where we need to increase the call stack, then lets do something. I Googled around for lodash templates callstack exceeded and didn't find much. Hopefully the issue you were running into was specific to the context in which lodash was being used.

kahnvex committed:
Completely remove handlebars from brwsrfy-i18n

This remove handlebars from browserify-i18n and sets up custom extension handelers for .handlebars and .hbs files.

beck commented:
Before trying to tackle handlebars (a lang that compiles to js), would it make sense to first handle translating javascript?

kahnjw commented:
>Before trying to tackle handlebars (a lang that compiles to js), would it make sense to first handle translating javascript?

Maybe, but we need to get this first pull in so we can translate sitelocation by the 24th. I was hoping this could move fairly fast.

kahnvex committed:
Always need to use custom extensions, more confident

kahnvex committed:
Whitespace

kahnvex committed:
translate function should be called i18n

beck commented on index.js:
```diff
@@ -0,0 +1,94 @@
+    var msgId = msgObject.msgid;
+    var msg = (msgObject.msgstr && msgObject.msgstr[0]) || msgId;
+    acc[msgId] = msg;
+  })
```
semicolon

beck commented on index.js:
```diff
@@ -0,0 +1,94 @@
+  return function(file, opts) {
+    var mergedOpts = {};
+
+    _.extend(mergedOpts, opts, fastOpts)
```
+;

kahnvex committed:
Missing semicolon

kahnvex committed:
Another missing semicolon

beck commented:
:+1:

Reservations I still hold:
* using lodash templates for handlebars makes me cringe
* there needs to be some separation between plugin code and filetype parsers
* the tests are a bit weak (would like to see some templates with nasty stings in them)
* travis is angry

euwest commented:
this isn't a ruby app and travis is trying to run rake tasks

kahnjw commented:
>using lodash templates for handlebars makes me cringe

I'm willing to wait and see on this one. It surprised me how much faster lodash templates are than Handlebars.

>there needs to be some separation between plugin code and filetype parsers

Yes. This needs to happen. We are duplicating code between gulp-static-i18n and here now. Lets consolidate that.

>the tests are a bit weak (would like to see some templates with nasty stings in them)

This I'll fix soon, maybe even before merging this.

>travis is angry

This I will fix before merging.

kahnjw commented:
>this isn't a ruby app and travis is trying to run rake tasks

I'm assuming we need to add a `.travis.yml` file. I think Travis is using its "sane" defaults.

kahnvex committed:
Add travis.yml config file

kahnvex committed:
Beef up the tests a little bit

These tests still need a lot of work, but this makes be a little more confident about the application code.

Issue: https://github.com/yola/browserify-i18n/issues/2

kahnvex committed:
Fix double quote escaping bug

euwest commented on README.md:
```diff
@@ -1,2 +1,70 @@
+var i18n = require('browserify-i18n');
+var opts = {
+  locale: 'en',                         // The locale code to be used
+  localeDirs: ['./path/to/locale/dir/'] // Array of paths to locale .po files
```
this `opts` object needs `global: true`

kahnvex replied:
Why?

euwest replied:
why does optimized have it and this one not?

kahnvex replied:
To illustrate which options must go to `i18n.fast` and which ones must go to the second argument of `b.transform`. We could use any option, but I think `global` is the only one you can give to a transform by default.

euwest replied:
in this case you're comparing optimized v. non-optimized and global v. local at the same time, which could be confusing to someone who's trying to figure out what this does. i'd suggest illustrating one feature at a time(in this case, optimized v. non-optimized, where the results would be exactly the same but one takes longer). i don't think we need to show that transform can take options, since we didn't write that functionality. but if you want to, i'd just make sure both examples are consistent
